### ISSUE TITLE
A performance issue of `vm_compute` with pattern matching returning a function

### ISSUE THREAD

**Issue by pi8027** [coq/coq#14109]
#### Description of the problem

Both of the following `Def1` and `Def2` module provides the factorial function. The only difference between them is that the `match` expressions in the definitions of `addn` and `muln` return `nat` in the former, and `nat -> nat` in the latter. This difference leads to a significant difference in performance.

```coq
Set Dump Bytecode.

Module Def1.

Fixpoint addn (n m : nat) : nat :=
  match n with
    | 0 => m
    | S n => S (addn n m)
  end.

Fixpoint muln (n m : nat) : nat :=
  match n with
    | 0 => 0
    | S n => addn m (muln n m)
  end.

Fixpoint factorial (n : nat) : nat :=
  match n with
    | 0 => 1
    | S n' => muln n (factorial n')
  end.

End Def1.

Module Def2.

Fixpoint addn (n : nat) : nat -> nat :=
  match n with
    | 0 => fun m => m
    | S n => fun m => S (addn n m)
  end.

Fixpoint muln (n : nat) : nat -> nat :=
  match n with
    | 0 => fun _ => 0
    | S n => fun m => addn m (muln n m)
  end.

Fixpoint factorial (n : nat) : nat :=
  match n with
    | 0 => 1
    | S n' => muln n (factorial n')
  end.

End Def2.

Time Eval vm_compute in Nat.even (Def1.factorial 11). (* Finished transaction in 2.064 secs (2.048u,0.015s) (successful) *)
Time Eval vm_compute in Nat.even (Def2.factorial 11). (* Finished transaction in 4.812 secs (4.727u,0.083s) (successful) *)
```

While propagating the arguments of a `match` expression into its branches is not possible by conversion, I hope there are some special cases allowing us to do this optimization in `vm_compute`, since `match` expression returning a function is useful in dependently typed programming. OCaml extraction has a similar issue and it is discussed in §4.3.4 of Pierre Letouzey's Ph.D. thesis.

#### Coq Version

8.13.2

**Comment by pi8027** [comment]
Reordering the arguments of a fixpoint function may also impact the performance significantly, but I don't understand what's going on:
```coq
Set Dump Bytecode.

Module Def3.

Fixpoint addn (n m : nat) : nat :=
  match m with
    | 0 => n
    | S m => addn (S n) m
  end.

Fixpoint muln (accu n m : nat) : nat :=
  match m with
    | 0 => accu
    | S m => muln (addn accu n) n m
  end.

Fixpoint factorial (n : nat) : nat :=
  match n with
    | 0 => 1
    | S n' => muln 0 (factorial n') n
  end.

End Def3.

Module Def4.

Fixpoint addn (m n : nat) : nat :=
  match m with
    | 0 => n
    | S m => addn m (S n)
  end.

Fixpoint muln (m n accu : nat) : nat :=
  match m with
    | 0 => accu
    | S m => muln m n (addn accu n)
  end.

Fixpoint factorial (n : nat) : nat :=
  match n with
    | 0 => 1
    | S n' => muln n (factorial n') 0
  end.

End Def4.

Time Eval vm_compute in Nat.even (Def3.factorial 11). (* Finished transaction in 2.004 secs (2.002u,0.s) (successful) *)
Time Eval vm_compute in Nat.even (Def4.factorial 11). (* Finished transaction in 13.905 secs (12.824u,1.047s) (successful) *)
```

**Comment by ppedrot** [comment]
> Reordering the arguments of a fixpoint function may also impact the performance significantly

Probably the VM manifestation of the phenomenon observed for native in https://github.com/coq/coq/issues/8019#issuecomment-403322104. Essentially, the accumulator check is not performed at the same place depending on the recursive argument.

> I hope there are some special cases allowing us to do this optimization in vm_compute

While I share the hope I also fear it's unlikely because we need to handle open terms via accumulators, which forbids *a lot* of possible reorderings. The few cases where we might optimize are when we statically know that a value can't be an accumulator, so maybe in nested matches (don't we do that already?) and similarly built fixpoints.

**Comment by silene** [comment]
> Reordering the arguments of a fixpoint function may also impact the performance significantly, but I don't understand what's going on

You are not computing the same thing in `Def3` or `Def4`, or rather, you are computing the same thing but in a much more complicated way. If you replace `Def4.muln` by
```coq
Fixpoint muln (m n accu : nat) : nat :=
  match m with
    | 0 => accu
    | S m => muln m n (addn n accu) (* <- change here *)
  end.
```
you will get much more sensible results.

**Comment by silene** [comment]
> Essentially, the accumulator check is not performed at the same place depending on the recursive argument.

No, the accumulator check is always performed at the same place, irrespective of the position of the recursive argument, as long as the fixpoint is fully applied, which is the case here.

**Comment by pi8027** [comment]
> You are not computing the same thing in `Def3` or `Def4`, or rather, you are computing the same thing but in a much more complicated way.

That's true. Sorry, my bad. I updated my comment.

**Comment by silene** [comment]
Be careful about what you are actually measuring. You might well be measuring the code of the garbage collector rather than the one of `vm_compute`. I suggest you run `Optimize Heap` before each call to `vm_compute`. By doing so, you should get sensibly the same results for `Def3` and `Def4`.

**Comment by silene** [comment]
As for the original issue with `Def1` and `Def2`, it might be possible to optimize it, despite accumulators. Indeed, since `fix` acts as a barrier for accumulators, the normal forms of `Def1.addn` and `Def2.addn` are actually identical (as long as one does not look at the actual code of the fixpoint). So, we could compile `Def2.addn` as `Def1.addn`, but when interpreting `GRABREC`, instead of pointing the accumulator closure "at the next `RESTART` instr", we could point it at the `RESTART` instruction of `Def2.addn`.

**Comment by pi8027** [comment]
> Be careful about what you are actually measuring. You might well be measuring the overhead of the garbage collector rather than `vm_compute`. I suggest you run `Optimize Heap` before each call to `vm_compute`. By doing so, you should get sensibly the same results for `Def3` and `Def4`.

That's also true. Thanks.

**Comment by silene** [comment]
> As for the original issue with `Def1` and `Def2`, it might be possible to optimize it, despite accumulators.

Now that I think again about it, there is a conceptually simpler change that does not require reasoning about accumulators and yet should apply in a lot more cases. For `Def2.addn` and `Def2.muln`, the compiler generates 4 times the sequence `CLOSURE;RETURN`. In other words, the interpreter will allocate on the heap and fill the block with all the free variables (`CLOSURE`), then check if there are some arguments on the stack and transfer control to the closure (`RETURN`), and then check if there are sufficiently many arguments on the stack (`GRAB`). So, there is a lot of wasted work.

If there was a `RETURNCLOSURE` opcode that combined all three operations, the interpreter would directly check if there are sufficiently many arguments on the stack and transfer control to the closure, without ever creating it on the heap. This should already bring a large speedup for a minimal change. Unfortunately, it will not work if there are free variables, since the code of the closure expects to find them on the heap (`ENVACC`), not on the stack (`ACC`).

So, the new opcode should rather be `BRANCHIFARGS` and the bytecode compiler would create two versions of the closures, one with free variables on the heap and one with free variables on the stack. Note that the version with free variables on the stack does not need any preprocessing work, because the free variables (e.g., arguments of the enclosing function) are already on the stack anyway. As a consequence, the only difference between the executions of `Def1.muln` and ` Def2.muln` would be the cheap `BRANCHIFARGS` opcode.

**Comment by ppedrot** [comment]
> the bytecode compiler would create two versions of the closures

Doesn't that mean we can get an exponential blowup when nesting matches?

**Comment by silene** [comment]
> Doesn't that mean we can get an exponential blowup when nesting matches?

You would need to nest an alternation of `match` and `fun`, all of them in terminal position. In that case, it would indeed cause an exponential blowup on bytecode size. But I don't think I have ever seen that kind of function in practice. (After all, there is just so many arrows you can put in your function signature, if you are not automatically generating it.) If it ever becomes an issue, we can apply a simple tweak: Do not perform this code duplication on the heap-closure path. So, only the fast inline path would get further optimizations.
